### DISCUSSION TITLE
Are closures not supported?

### DISCUSSION THREAD

**Discussion by mmontone** [hylang/hy#2583] [Q&A] [answered]
Hi. Are closures not supported?

The following code all gives: `UnboundLocalError: local variable 'x' referenced before assignment`:

Reading a variable from closure context:
```
(setv my-closure
      (let [x 0]
        (fn []
          x)))

(my-closure)
```
=> `UnboundLocalError: local variable 'x' referenced before assignment`

Setting a variable in closure context:
```
(setv my-closure
      (let [x 0]
        (fn []
          (setv x (+ x 1))
          x)))

(my-closure)
```
=> `UnboundLocalError: local variable 'x' referenced before assignment`

**Comment by Kodiologist** [comment, marked as answer]
Yes, closures are supported. Any Python code using closures will work the same if translated into Hy.

Your first example doesn't work because there's no built-in macro named `setf`. Note that the error message is actually `NameError: name 'setf' is not defined`, not `UnboundLocalError`.

Your second example doesn't work because your function tries to set `x`, which means it will create a new local variable for `x`, which is unset by the time of `(+ x 1)`. Use `(nonlocal x)` to avoid this. In general, I'd recommend mastering Python scoping before trying to use `let`, which complicates things.

**Comment by mmontone** [comment]
Ok. Thanks. I'll have a look.

What do you mean with the `setf` in the first example? I'm just trying to read a variable introduced with the `let`, there's not variable setting.

**Reply by mmontone** [reply]
Sorry. I changed the code there. I mis typed.

**Comment by mmontone** [comment]
It works after `nonlocal`! Thanks!
